Add support for ephemeral arguments in variables

Rishang opened the pull request:
Fix https://github.com/shihanng/tfvar/issues/38

Commit:
Add support for ephemeral variables in module configuration

Commit:
Update release workflow to use --clean argument for goreleaser

Commit:
Update GitHub Actions workflows to use updated action versions for Go, checkout, and golangci-lint

Commit:
Update goreleaser configuration to skip Homebrew tap and adjust golangci-lint version in workflows

Commit:
Refactor WriteAsTFEResource to conditionally set ephemeral attribute for variables

Commit:
Update goreleaser configuration to version 2

Commit:
Remove Homebrew tap configuration and disable changelog generation in goreleaser settings

Commit:
Add Homebrew tap configuration for tfvar with installation and test instructions

shihanng commented on pkg/tfvar/tfvar.go:
```diff
@@ -178,6 +180,9 @@ func WriteAsTFEResource(w io.Writer, vars []Variable) error {
 		resourceBody.SetAttributeValue("sensitive", cty.BoolVal(v.Sensitive))
+		if v.Ephemeral {
+			resourceBody.SetAttributeValue("ephemeral", cty.BoolVal(v.Ephemeral))
+		}
```
Are you sure we need this? I don't see the attribute in `tfe_variable` documentation: https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/resources/variable

shihanng commented on pkg/tfvar/testdata/ephemeral/main.tf:
```diff
@@ -0,0 +1,23 @@
+  type      = string
+  default   = "temp-value"
+  ephemeral = true
+}
```
This file is not being tested?

I suggest only add to `pkg/tfvar/testdata/defaults/main.tf` and fix the test accordingly.

```
variable "session_id" {
  type      = string
  ephemeral = true
}
```

shihanng reviewed (commented):
Thanks for the PR. I added a few comments. Since, they are unrelated, please remove all changes regarding CI from this PR. This reduces noise and allow us to easily rollback in case of needed.

shihanng commented:
Hi @Rishang, do you still plan to work on this PR?

Commit:
Add support for ephemeral variables in tests and configuration files

Rishang commented:
@shihanng  I have made few changes now

Rishang commented:
Please ignore github workflows I will revert those once you approve changes

Commit:
fix: ephemeral variable support in workspace payload test data

shihanng commented:
> Please ignore github workflows I will revert those once you approve changes

I can't approve the PR (it does not make sense to approve changes that I disagree with).
Please remove all changes regarding CI from this PR.

Feel free to drop the commits, changing the history of this branch. Thank you.

Commit:
fix: revert ci changes

Rishang commented:
I have reverted CI changes

Rishang commented:
Do note that the CI plugins you are using are depricated so i was testing to fix that but no having much idea in CI so i have reverted it

shihanng commented on pkg/tfvar/tfvar_test.go:
```diff
@@ -54,6 +54,20 @@ func TestLoad(t *testing.T) {
+	}
+	assert.True(t, found, "expected to find session_id variable")
+}
+
```
Why do we need this test case? It this not already covered in `TestWriteAsEnvVars`?

Rishang replied:
I have removed this test field

shihanng commented on pkg/tfvar/tfvar.go:
```diff
@@ -15,14 +15,16 @@ import (
+//
+//	variable "image_id" {
+//	  type = string
+//	}
```
Please revert this change. It is unrelated to the change we want to address.

shihanng commented on pkg/tfvar/tfvar.go:
```diff
@@ -55,7 +58,8 @@ func Load(dir string) ([]Variable, error) {
 // WriteAsEnvVars outputs the given vars in environment variables format, e.g.
-//    export TF_VAR_region='ap-northeast-1'
+//
+//	export TF_VAR_region='ap-northeast-1'
```
Ditto

shihanng commented on pkg/tfvar/tfvar.go:
```diff
@@ -100,7 +104,8 @@ func oneliner(original hclwrite.Tokens) hclwrite.Tokens {
 // WriteAsTFVars outputs the given vars in Terraform's variable definitions format, e.g.
-//    region = "ap-northeast-1"
+//
+//	region = "ap-northeast-1"
```
Ditto

shihanng commented on pkg/tfvar/tfvar.go:
```diff
@@ -172,12 +177,16 @@ func WriteAsTFEResource(w io.Writer, vars []Variable) error {
 		resourceBody.SetAttributeValue("key", cty.StringVal(v.Name))
 		resourceBody.SetAttributeValue("value", v.Value)
 		resourceBody.SetAttributeValue("sensitive", cty.BoolVal(v.Sensitive))
+
```
Ditto

shihanng commented on pkg/tfvar/tfvar.go:
```diff
@@ -172,12 +177,16 @@ func WriteAsTFEResource(w io.Writer, vars []Variable) error {
 	for _, v := range vars {
+		if v.Ephemeral {
+			continue
+		}
```
Why do we need to skip when it is ephemeral? Can you share your thought process here or any documentations that support this?

Rishang replied:
We skip it as tfe_variable do not have any ephemeral resource like you mentioned previously https://github.com/shihanng/tfvar/pull/39#discussion_r2553455591. ephemeral vars are meant to be transient and never saved to state via any terraaform resource

shihanng replied:
Although we do not have the `ephemeral` attribute (therefore could not set it), I think it makes sense to create the resource. The resource is for setting the value of the Terraform variable. The user will have the freedom to not use it.

I suggest we not skip it. Does it make sense?

Rishang replied:
I think we should ingore the argument if its not present to not make any confusion

shihanng replied:
@Rishang here we are skipping the whole variable just because it has the `ephemeral` attribute. I don't think this makes sense.

Rishang replied:
I have removed skip condition

shihanng commented on pkg/tfvar/testdata/defaults/main.tf:
```diff
@@ -61,3 +61,14 @@ variable "with_optional_attribute" {
+  type      = string
+  ephemeral = true
+  default = "some-ephemeral-value"
+}
```
Please remove this block. We do not need to test for default values because that is already covered in other examples. For example, in `variable "password"` we also not test for `default`.

Rishang replied:
removed

```
variable "ephemeral_variable" {
  type      = string
  ephemeral = true
  default = "some-ephemeral-value"
}
```

shihanng reviewed (commented):
I added some comments.

Please ignore the typecheck comments from CI. I will work on the CI fix.

Commit:
fix: fmt

Commit:
Remove unused ephemeral variable from main.tf

Commit:
Remove TestLoad_Ephemeral and associated ephemeral variable from tests and workspace payload

shihanng commented:
Some tests (not linting) are failing. We should fix that too.

Commit:
refactor: adjust formatting of variable assignments in tfvar_test.go

Commit:
remove handling of ephemeral variables in WriteAsTFEResource function

Rishang committed:
Merge branch 'trunk' into trunk

Commit:
Add support for ephemeral variables in WriteAsTFEResource function and update test data

- Implemented handling of ephemeral attribute in the WriteAsTFEResource function.
- Added a new test case for an ephemeral variable in the test data file.

Rishang commented:
Please check now

shihanng commented:
@Rishang regarding your last commit d696809. Why did you add the `ephemeral` attribute back to `tfe_variable` resource? We had the discussion in https://github.com/shihanng/tfvar/pull/39#discussion_r2553455591. I still don't see the `ephemeral` in the [resource](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/resources/variable#attributes-reference). Am I missing something here?

shihanng commented on pkg/tfvar/tfvar.go:
```diff
@@ -181,6 +183,9 @@ func WriteAsTFEResource(w io.Writer, vars []Variable) error {
 		resourceBody.SetAttributeValue("category", cty.StringVal("terraform"))
+		if v.Ephemeral {
+			resourceBody.SetAttributeValue("ephemeral", cty.BoolVal(true))
+		}
```
Please remove this then run
```
go test  ./... -update
```
to update the golden files.

See [comment here](https://github.com/shihanng/tfvar/pull/39#issuecomment-3951067036).

Rishang committed:
fix: remove ephemeral argument

rishangbhavsar commented:
I have removed it now
